Forward lines from stderr of child process to stdout on the same thread, instead of spawning a thread

dpaoliello opened the pull request:
`cc-rs` was redirecting output from `stderr` of its child processes to a pipe, which was then being read on dedicated thread and written to `stdout` with a wrapper message.

This pattern caused a few issues:
* It created a dependency from `cc-rs` to `libc` when building the Rust standard library with optimized compiler intrinsics, resulting in the Rust build breaking (see #913).
* With the `parallel` feature enabled, there was no synchronization between different child processes writing to the pipe, potentially resulting in corrupted output.
* The thread that was forwarding the lines was wasted, since the main thread was blocking waiting for the child process(es) to exit anyway.

The fix for this is to forward output from `stderr` to `stdout` on the current thread. For the `parallel` feature this required implementing a non-blocking version of `BufReader::read_until` that consumes whatever data is available and then continues.

Fixes #913

NobodyXu commented:
There're a lot of changes here, it might take me some time to review it.

dpaoliello commented:
If it helps, I have a build of the Rust toolchain that uses this: https://github.com/rust-lang/rust/actions/runs/7719025147/job/21041489650?pr=119199

NobodyXu commented on src/command_helpers.rs:
```diff
@@ -41,83 +40,55 @@ impl CargoOutput {
-    pub(crate) fn new() -> Result<Self, Error> {
-        let (pipe_reader, pipe_writer) = crate::os_pipe::pipe()?;
+    #[cfg(feature = "parallel")]
+    pub(crate) fn reader_for_stderr_as_warnings(
```
I think it doesn't have to be an associated fn of CargoOutput, child.stdout() already has the information required.

NobodyXu commented on src/command_helpers.rs:
```diff
@@ -41,83 +40,55 @@ impl CargoOutput {
-        // location to another.
-        let print = thread::spawn(move || {
-            let mut stderr = BufReader::with_capacity(4096, pipe_reader);
+    fn forward_stderr_as_warnings(&self, child: &mut Child) {
```
Same here, it doesn't have to be an associated fn.

NobodyXu commented on src/command_helpers.rs:
```diff
@@ -307,8 +263,27 @@ pub(crate) fn try_wait_on_child(
+        if let Ok(mut available) = child_stderr_reader.fill_buf() {
+            let stdout = io::stdout();
+            let original_size = available.len();
+            while let Some(i) = memchr::memchr(b'\n', available) {
```
I'd prefer not to add dep `memchr`.

NobodyXu commented on src/command_helpers.rs:
```diff
@@ -307,8 +263,27 @@ pub(crate) fn try_wait_on_child(
+    // between the child printing messages and then exiting.
+    let wait_result = child.try_wait();
+
+    if let Some(child_stderr_reader) = child_stderr_reader.as_mut() {
```
I'm thinking if it makes more sense to do this in the async runtime, i.e. if we add poll/select/epoll/waitfor_multiple_objects in the future to speed it up, then it makes sense to do it there.

However since it'd be done in the future, you don't have to change this code.

Some refactoring to extract a new fn is enough.

NobodyXu commented on src/command_helpers.rs:
```diff
@@ -41,83 +40,51 @@ impl CargoOutput {
+    child
+        .stderr
+        .take()
+        .map(|stderr| BufReader::with_capacity(4096, stderr))
```
I feel like we don't need 4096 B for the buffer, IMO I rarely see any warning for my project using cc-rs indirectly, and even if it does it's unlikely to print 4096B of characters.

If that happens, it will likely be a compilation error in C++ which dumps a lot of messages due to overloading.

It's so rare and usually signals a failure, so I don't think we need to optimize for that particular use case.

We should instead tune it down to save memory, allocating a whole page is going to be quite expensive since it's large.

I'd recommend something like 100B, should be good enough.

dpaoliello replied:
I'll switch this to `BufReader::new` instead of explicitly setting a capacity.

NobodyXu commented on src/command_helpers.rs:
```diff
@@ -41,83 +40,51 @@ impl CargoOutput {
-                    stdout.write_all(b"\n").unwrap();
-                }
+fn forward_all_stderr_as_warnings(child: &mut Child) {
+    if let Some(stderr) = child.stderr.as_mut() {
```
Let's use `reader_for_stderr_as_warning` here, to reuse buffer configuration.

NobodyXu commented on src/command_helpers.rs:
```diff
@@ -41,83 +40,51 @@ impl CargoOutput {
 
-        self.handle.take().unwrap().join().unwrap();
-    }
+    stdout.write_all(b"cargo:warning=").unwrap();
```
Shame that we don't have write_all_vectored

NobodyXu commented on src/command_helpers.rs:
```diff
@@ -301,14 +253,38 @@ pub(crate) fn command_add_output_file(
+        if let Ok(available) = child_stderr_reader.fill_buf() {
+            let stdout = io::stdout();
+            let mut consumed = 0;
+            for line in available.split(|&b| b == b'\n') {
```
I think we need to handle cases where `available` does not include a newline and there's more data to come.

In this case, I think you should keep the data until newline or eof, because if you prints part of the line as a warning, then you would be adding newline to the originaly message.

dpaoliello replied:
Ah, yep, I forgot that `split` will provide the rest of the items in last iteration.

NobodyXu commented on src/command_helpers.rs:
```diff
@@ -301,14 +248,41 @@ pub(crate) fn command_add_output_file(
+            let mut consumed = 0;
+            for line in available.split_inclusive(|&b| b == b'\n') {
+                // Only forward complete lines, leave the rest in the buffer.
+                if let Some((b'\n', line)) = line.split_last() {
```
Isn't it more efficient to check the last byte of the buffer directly?

dpaoliello replied:
`write_warning` doesn't expect the line to end with `\n`, so I'd have to trim it off anyway. Looking at the implementation for `split_last`, it doesn't seem that expensive.

NobodyXu replied:
Thanks, I misunderstood the API of split_last

NobodyXu commented on src/command_helpers.rs:
```diff
@@ -301,14 +248,41 @@ pub(crate) fn command_add_output_file(
+                // Only forward complete lines, leave the rest in the buffer.
+                if let Some((b'\n', line)) = line.split_last() {
+                    consumed += line.len() + 1;
+                    write_warning(&stdout, line);
```
I think you also need to handle incomplete line on eof.

Tip: You could convert this function to an async one to make it easier to write, compile_objects does have some async support if you follow the example

dpaoliello replied:
Ugh, I think the only way that I can handle that is to use `forward_all_stderr_as_warnings` once the process has exited, otherwise there's no way with `BufReader` to know if the underlying data source still has data left or has hit EOF without consuming all the bytes currently in the buffer.

dpaoliello replied:
I also don't know how async would help here...

NobodyXu replied:
async would make handling internal state easier, so instead of using BufReader you can just set to childstderr to non-blocking, keep reading from the child into a Vec and find newline until it returns `EWOULDBLOCK`, then yield control until it resumes.

dpaoliello replied:
I don't see how to do this with the API exposed by `ChildStdErr`, unless you mean to convert it into its underlying fd or handle, which doesn't seem portable.

NobodyXu replied:
That's true...it's only possible if you create the pipe manually.

Perhaps just add a detection of eof and then flush all pending data in the buffer?

dpaoliello commented on src/command_helpers.rs:
```diff
@@ -41,83 +40,46 @@ impl CargoOutput {
+pub(crate) fn reader_for_stderr_as_warnings(
+    child: &mut Child,
+) -> Option<BufReader<std::process::ChildStderr>> {
+    child.stderr.take().map(BufReader::new)
```
Ok, back to 100b buffer.

The `BufReader` makes things more convenient, otherwise if we were using the `ChildStderr` directly we'd need to shuffle any remaining bytes after the final `\n` back to the beginning of the `Vec` between each `read` OR we would have to `read_to_end` which could result in a very large `Vec`.

NobodyXu replied:
> otherwise if we were using the ChildStderr directly we'd need to shuffle any remaining bytes after the final \n back 

Hmm I think this isn't very hard with [`Vec::drain`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.drain), just two lines for splitting based on newline and remove it using drain after printing.

dpaoliello replied:
This feels like a premature optimization: we'd have to choose a size for the `Vec` to read into, which is no different to the buffer in the `BufReader`. It would avoid the second allocation, but this is in the context of spawning and waiting for a process so the execution time is dominated by that rather than the work here. I'd prefer to keep the code simple than make the optimization to avoid a `BufReader`. 

We've already made improvements: 100b for the `BufReader` instead of 4kb and no allocations for the `Vec` if there's no warnings.

NobodyXu replied:
> It would avoid the second allocation, but this is in the context of spawning and waiting for a process so the execution time is dominated by that rather than the work here. I'd prefer to keep the code simple than make the optimization to avoid a `BufReader`. 

Understood, that is indeed the case.

> We've already made improvements: 100b for the `BufReader` instead of 4kb and no allocations for the `Vec` if there's no warnings. 

Yeah that's probably good enough.

Thanks.

dpaoliello commented:
@NobodyXu I'm going to have to rework this, since there are two issues:

The first, fixable, issue is that `BufReader.fill_buf` will only `read` more if all data in the buffer has been consumed - so if we don't consume everything then `fill_buf` won't read any more data and so each subsequent call to `fill_buf` will keep returning the same data. 

The second, fundamental, issue is that `read` will block until data is available or the child exits - so in parallel mode we'll block waiting for the first child's `stderr` to return something or until that process exits.

There's a few different ways that we can address these:
1. Accept the limitations of the second issue, and workaround the first issue by replacing the `BufReader` with our own calls to `read` into a `Vec` (as you had suggested for the non-parallel code).
2. Give up on trying to forward the `stderr` output while the children are running, and instead forward them all once the child has exited.
3. Do something like the original code but create one thread per child instead of using a single thread.

Thoughts?

NobodyXu commented:
> The second, fundamental, issue is that read will block until data is available or the child exits - so in parallel mode we'll block waiting for the first child's stderr to return something or until that process exits.

If you bring back os_pipe, then you could read without blocking.

dpaoliello commented:
@NobodyXu ok, I think I have this figured out: for Unix, I'm setting the pipe as non-blocking and checking for `WouldBlock`. For Windows, I'm using `PeekNamedPipe` to check for data and returning early if there is none.

I also removed the `BufReader` and switched to directly reading into the `Vec`.

NobodyXu commented on src/windows/windows_sys.rs:
```diff
@@ -55,16 +55,18 @@ extern "system" {
-        lppipeattributes: *const SECURITY_ATTRIBUTES,
-        nsize: u32,
-    ) -> BOOL;
+    pub fn OpenSemaphoreA(dwdesiredaccess: u32, binherithandle: BOOL, lpname: PCSTR) -> HANDLE;
```
Please update https://github.com/rust-lang/cc-rs/blob/main/dev-tools/gen-windows-sys-binding/windows_sys.list and then run the crate to generate this file, this file is not meant for modification directly.

dpaoliello replied:
Ah, I was wondering where that file was - thanks!

NobodyXu commented on src/command_helpers.rs:
```diff
@@ -41,83 +40,175 @@ impl CargoOutput {
+            #[cfg(all(windows, feature = "parallel"))]
+            let is_non_blocking = self.is_non_blocking;
+            let mut read_stderr = move |buf: &mut [u8]| -> Result<usize, io::Error> {
+                // On Unix, the pipe is non-blocking, so we can just read.
```
You need to set the pipe to non-blocking on unix, it's not non-blocking by default.

dpaoliello replied:
It's set to non-blocking by `set_non_blocking` - I've updated the comment to match and added a debug assert to check the flags.

NobodyXu commented on src/command_helpers.rs:
```diff
@@ -41,83 +40,175 @@ impl CargoOutput {
-            let mut stderr = BufReader::with_capacity(4096, pipe_reader);
-            let mut line = Vec::with_capacity(20);
-            let stdout = io::stdout();
+#[cfg(all(feature = "parallel", not(unix), not(windows)))]
```
At this point, why not bring back os_pipe mod and move the unsafe function there?

dpaoliello replied:
I'll move these unsafe functions into their own module under `parallel`, but there's no need to resurrect the rest of `os_pipe`.

NobodyXu commented on src/parallel/stderr.rs:
```diff
@@ -0,0 +1,84 @@
+        let fd = stderr.as_raw_fd();
+        let flags = get_flags(&fd)?;
+
+        if unsafe { libc::fcntl(fd, libc::F_SETFL, flags | libc::O_NONBLOCK) } != 0 {
```
Honestly I think you don't need get_flags().

From the [man page of fcntl](https://man7.org/linux/man-pages/man2/fcntl.2.html)

>  On Linux, this command can change
>               only the O_APPEND, O_ASYNC, O_DIRECT, O_NOATIME, and
>               O_NONBLOCK flags.  It is not possible to change the
>               O_DSYNC and O_SYNC flags; see BUGS, below.

It's a pipe on unix, so only O_NONBLOCK makes sense IMHO.

dpaoliello replied:
Fair enough - I've added debug asserts just to be safe.

NobodyXu commented on src/parallel/stderr.rs:
```diff
@@ -0,0 +1,84 @@
+            }
+        }
+        if bytes_available == 0 {
+            return Err(std::io::Error::new(
```
Recommend  to use <io::Error as From> impl for io::ErrorKind, io::Error::new allocates internally while the from errorkind does not.

NobodyXu commented on src/command_helpers.rs:
```diff
@@ -41,83 +40,122 @@ impl CargoOutput {
+                    }
+                    Err(err) if err.kind() == std::io::ErrorKind::Interrupted => {
+                        // Interrupted, try again.
+                        buffer.truncate(old_data_end);
```
resize is honestly a bit expensive as it have to set the rest of the buffer to 0.

You can either:
 - move the resize outside the loop, and make sure every match returns.
 - use `MaybeUninit`, but given that the `std::io::Read` trait only takes `&[u8]`, it in theory can be a UB if the implementator of `Read` actually reads from it, instead of just writing to it.

dpaoliello replied:
I switched this to `set_len` since I'm always truncating to the initialized data after reading - I can't move the `resize` outside of the loop (in case we've filled the buffer but haven't seen a `\n`), and I don't think that `MaybeUninit` supports dynamic sized buffers.

NobodyXu replied:
I was thinking of `Vec<MaybeUninit<u8>>`

dpaoliello replied:
I don't think that works, `MaybeUninit` isn't guaranteed to be the same layout as the underlying type: https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#layout-1 so you can't get a `&[u8]` from a `Vec<MaybeUninit<u8>>`.

NobodyXu replied:
Thanjs, I didn't know that.

NobodyXu commented on src/command_helpers.rs:
```diff
@@ -41,83 +40,122 @@ impl CargoOutput {
+                        if old_data_end > 0 {
+                            write_warning(&buffer[..old_data_end]);
+                        }
+                        return true;
```
Let's set `self.inner = 0;` here, so that once `true` is returned, it will always returns `true` in the future (because in parallel `compile_objects`, it can be called after `true` is returned).

dpaoliello replied:
Good call.

dpaoliello commented on src/command_helpers.rs:
```diff
@@ -41,83 +40,122 @@ impl CargoOutput {
+        if let Some((stderr, buffer)) = self.inner.as_mut() {
+            fn write_warning(line: &[u8]) {
+                let stdout = io::stdout();
+                let mut stdout = stdout.lock();
```
Unfortunately this causes errors with older versions of Rust: https://github.com/rust-lang/cc-rs/actions/runs/7764282806/job/21177496391

dpaoliello commented on src/command_helpers.rs:
```diff
@@ -41,83 +40,122 @@ impl CargoOutput {
+            let read_stderr = ChildStderr::read;
+
+            loop {
+                buffer.reserve(MIN_BUFFER_CAPACITY);
```
If the buffer is full but we haven't seen a `\n` yet, then we need to expand it to read more to find that `\n` (since we can't write a partial line).

NobodyXu commented on src/parallel/stderr.rs:
```diff
@@ -0,0 +1,74 @@
+                return Err(std::io::Error::last_os_error());
+            }
+        }
+        if bytes_available == 0 {
```
I wonder if it makes sense to pass the `Vec` here, then `.reserve(bytes_available)` to make sure all remaining data is read.

dpaoliello replied:
With the change to use `FIONREAD`, I've modified the code to split checking available data from doing the actual read, so it's pretty easy to reserve at least the bytes currently available.

NobodyXu commented on src/parallel/stderr.rs:
```diff
@@ -0,0 +1,74 @@
+    // On Unix, the pipe set to non-blocking by set_non_blocking, so we can just read.
+    #[cfg(unix)]
+    {
+        use std::os::unix::io::AsRawFd;
```
Nit: ioctl supports FIONREAD on non-blocking pipe, to retrieve data left for reading.

dpaoliello replied:
Oh, nice! this really simplifies the code since Windows and Unix can then use the same pattern. Thanks!

NobodyXu reviewed (changes requested):
Thanks, it overall LGTM, just two nitpicks I want to discuss with you before I approve and merge it.

NobodyXu commented on src/command_helpers.rs:
```diff
@@ -41,83 +40,110 @@ impl CargoOutput {
-    pub(crate) fn clone_pipe_writer(&self) -> Result<File, Error> {
-        self.try_clone_pipe_writer().map(Option::unwrap)
+impl StderrForwarder {
+    pub(crate) fn new(child: &mut Child) -> Self {
```
You still need to set the childstderr as non-blocking.

I read tht FIONREAD only works on pipe if it's non-blocking.

Also, I want to play safe and always use non-blocking to prevent deadlock.

dpaoliello replied:
Ok, I resurrected this non-blocking code.

NobodyXu commented on src/parallel/stderr.rs:
```diff
@@ -0,0 +1,47 @@
+    #[cfg(unix)]
+    {
+        use std::os::unix::io::AsRawFd;
+        if unsafe { libc::ioctl(stderr.as_raw_fd(), libc::FIONREAD, &mut bytes_available) } != 0 {
```
Stackoverflow says some implementation is broken and fail none-the-less.

I'd like it to fallback to reserving 100B on failure, and have an atomic for checking is the impl is buggy/supported.

dpaoliello replied:
Ugh, ok.

I'm also noticing that I'm getting spurious 0-bytes available on Linux (Ubuntu 20.04 in WSL) as well...

dpaoliello replied:
I'm also not a fan of disabling this across the entire process after a single pipe fails, so I've added a per-pipe flag if it fails or returns 0.

NobodyXu commented on src/parallel/stderr.rs:
```diff
@@ -0,0 +1,75 @@
+            ));
+        }
+    }
+    Ok(bytes_available as usize)
```
I don't like `as` casting, please use `.try_into().unwrap()` so that it would fail if casting negative number to `usize`, or when casting number too large to fit in `usize`

NobodyXu commented:
Thanj you!

dpaoliello commented:
Can we please get a release with this change?

NobodyXu commented:
Yes, I definitely want to have a release.

There's a regression #902 that needs to be fixed before a release can be done.
